dali/kernels/slice/slice_flip_normalize_permute_pad_cuda_impl.cuh

```diff
@@ -31,122 +32,153 @@
+  int channel_dim;
+  bool need_pad;
+  bool need_flip;
+
+  fast_div<uint64_t> out_strides[Dims];
+  TensorShape<Dims> in_strides;
 };
 
 struct BlockDesc {
   int sampleIdx;
-  size_t offset;
-  size_t size;
+  uint64_t offset;
+  uint64_t size;
 };
 
-template <typename OutputType, typename InputType, unsigned Dims, bool should_normalize>
-__device__ inline void SliceFlipNormalizePermutePadFunc(OutputType *__restrict__ out,
-                                                        const InputType *__restrict__ in,
-                                                        const int64_t *out_strides,
-                                                        const int64_t *in_strides,
-                                                        const int64_t *out_shape,
-                                                        const int64_t *padded_out_shape,
-                                                        bool should_pad,
-                                                        unsigned norm_dim,
-                                                        const float *norm_add,
-                                                        const float *norm_mul,
-                                                        OutputType padding_val,
-                                                        size_t offset, size_t block_end) {
-  if (Dims > 1 && !should_normalize &&
-      out_strides[Dims - 1] == in_strides[Dims - 1] &&
-      out_shape[Dims - 1] == padded_out_shape[Dims - 1]) {
-    const unsigned NextDims = Dims > 1 ? Dims - 1 : 1;
-    SliceFlipNormalizePermutePadFunc<OutputType, InputType, NextDims, should_normalize>(
-        out, in, out_strides, in_strides, out_shape, padded_out_shape,
-        should_pad, norm_dim, norm_add, norm_mul, padding_val, offset, block_end);
+DALI_HOST_DEV DALI_FORCEINLINE bool is_out_of_bounds(int64_t idx, int64_t data_extent) {
+  // check idx < 0 and idx >= data_extent at once
+  return static_cast<uint64_t>(idx) >= static_cast<uint64_t>(data_extent);
+}
+
+/**
+ * @brief General algorithm that allows for padding in any dimension
+ * @remarks `in` refers to the slice anchor start
+ */
+template <bool NeedFlip, bool NeedNormalize, bool NeedPad, int Dims,
+          typename Out, typename In, bool AllDims = true>
+__device__ void SliceFlipNormalizePermutePadFunc(
+    Out *__restrict__ out, const In *__restrict__ in,
+    const fast_div<uint64_t> *out_strides, const int64_t *in_strides, const int64_t *out_shape,
+    const int64_t *in_shape, const int64_t *anchor, const Out *__restrict__ fill_values,
+    const float *__restrict__ norm_add, const float *__restrict__ norm_mul, int channel_dim,
+    uint64_t offset, uint64_t block_end) {
+  if (Dims > 1 && out_strides[Dims - 1] == in_strides[Dims - 1] && anchor[Dims - 1] == 0 &&
+      out_shape[Dims - 1] == in_shape[Dims - 1] && channel_dim != Dims - 1 &&
+      (!NeedFlip || (in_strides[Dims - 1] < 0) == (in_strides[Dims - 2] < 0))) {
```

Comment:
When you can have negative strides?

Reply:
flipping. Anchor is at the end of the slice and the strides are negative